lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -2,271 +2,535 @@
+  auto *tuple = cast<TupleInst>(operand->getUser());
+  Operand *singleUse = tuple->getSingleUse();
+  if (!singleUse || !isa<ReturnInst>(singleUse->getUser()))
+    return tuple;
+
+  SILFunction *function = tuple->getFunction();
+  if (function->getConventions().getNumDirectSILResults() < 2)
+    return tuple;
+
+  unsigned resultIdx = tuple->getElementIndex(operand);
+
+  auto loweredFnConv = getLoweredFnConv(function);
+  assert(loweredFnConv.getResults().size() == tuple->getElements().size());
+
+  unsigned indirectResultIdx = 0;
+  for (SILResultInfo result : loweredFnConv.getResults().slice(0, resultIdx)) {
+    if (loweredFnConv.isSILIndirect(result))
+      ++indirectResultIdx;
+  }
+  // Cannot call F->getIndirectSILResults here because that API uses the
+  // function conventions before address lowering.
+  return function->getArguments()[indirectResultIdx];
+}
 
-  ValueVector::iterator end() { return valueVector.end(); }
+/// Return the value representing storage for a single return value.
+///
+///   bb0(%loweredIndirectResult : $*T, ...) // function entry
+///     return %oper
+///
+///   For %oper, return %loweredIndirectResult
+static SILValue getSingleReturnAddress(Operand *operand) {
+  assert(!isPseudoReturnValue(operand->get()));
+
+  auto *function = operand->getParentFunction();
+  auto loweredFnConv = getLoweredFnConv(function);
+  assert(loweredFnConv.getNumIndirectSILResults() == 1);
+  (void)loweredFnConv;
+
+  // Cannot call getIndirectSILResults here because that API uses the
+  // function conventions before address lowering.
+  return function->getArguments()[0];
+}
 
-  ValueVector::reverse_iterator rbegin() { return valueVector.rbegin(); }
+//===----------------------------------------------------------------------===//
+//                              ValueStorageMap
+//
+//              Map Opaque SILValues to abstract storage units.
+//===----------------------------------------------------------------------===//
 
-  ValueVector::reverse_iterator rend() { return valueVector.rend(); }
+/// Check if this is a copy->store pair. If so, the copy storage will be
+/// projected from the source, and the copy semantics will be handled by
+/// UseRewriter::visitStoreInst.
+static bool isStoreCopy(SILValue value) {
+  auto *copyInst = dyn_cast<CopyValueInst>(value);
+  if (!copyInst)
+    return false;
 
-  bool contains(SILValue value) const {
-    return valueHashMap.find(value) != valueHashMap.end();
-  }
+  if (!copyInst->hasOneUse())
+    return false;
 
-  unsigned getOrdinal(SILValue value) {
-    auto hashIter = valueHashMap.find(value);
-    assert(hashIter != valueHashMap.end() && "Missing SILValue");
-    return hashIter->second;
-  }
+  auto *user = value->getSingleUse()->getUser();
+  return isa<StoreInst>(user) || isa<AssignInst>(user);
+}
 
-  ValueStorage &getStorage(SILValue value) {
-    return valueVector[getOrdinal(value)].second;
-  }
+ValueStorage &ValueStorageMap::insertValue(SILValue value) {
```

Comment:
`insertValue` should not return a reference to the key, but receive the mapped value(s) as a parameter: either a `ValueStorage` or just a SILValue (the `storageAddress`

Reply:
> insertValue should not return a reference to the key, but receive the mapped value(s) 
That makes sense now. It's was more flexible the old way, but no longer important.